### ISSUE TITLE
Invalid build config for Org.WebRtc.csproj

### ISSUE THREAD

**Issue by djee-ms** [webrtc-uwp/webrtc-uwp-sdk#150]
Hi,

The currently checked-in C# project for `Org.WebRtc` contains an invalid platform configuration which builds x86 on Debug and x64 on Release. This generates some mismatching configuration with a Debug x64 build of the native C wrappers resulting in subtle bugs where _e.g._ the ICE configuration is parsed with invalid offsets to fields, resulting in wrong values that may or may not fail the config validation and the peer connection creation.

From [`projects/msvc/Org.WebRtc.NetStandard/Org.WebRtc.csproj`](https://github.com/webrtc-uwp/webrtc-windows/blob/releases/m71/projects/msvc/Org.WebRtc.NetStandard/Org.WebRtc.csproj):
```
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
    <PlatformTarget>x86</PlatformTarget>
    <OutputPath>$(SolutionDir)Build\Output\Net\$(MSBuildProjectName)\$(Configuration)\$(Platform)\</OutputPath>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
    <PlatformTarget>x64</PlatformTarget>
    <OutputPath>$(SolutionDir)Build\Output\Net\$(MSBuildProjectName)\$(Configuration)\$(Platform)\</OutputPath>
</PropertyGroup>
```

**Comment by djee-ms** [comment]
This doesn't seem to be enough. The code generated for calling:
```cpp
absl::optional<int> RTCConfigurationToIceConfigOptionalInt(int rtc_configuration_parameter)
```
from `PeerConnection::ParseIceConfig()` in `webrtc\xplatform\webrtc\pc\peerconnection.cc` reads `config.ice_connection_receiving_timeout` at offset 0x84 from the beginning of the `PeerConnectionInterface::RTCConfiguration` structure. However in the Visual Studio debugger this offset is reported by the Watch window as being 0x7C. Basically, the Visual Studio debugger shows a value in the caller (extracted from the debugger reading the `config` structure), then a different value for the `rtc_configuration_parameter` argument in the callee because the machine code actually reads at the wrong offset.

I am wondering if the code generated for the wrapper (by ZsLib?) was not compiled with different options (_e.g._ compiled by Clang, or compiled by MSVC x86) than the ones it's being used with (MSVC x64), because this structure has things like `std::vector` which depend on the `std` library used as well as the architecture. I see that `pc/peerconnection` is built by the `win:x64` toolchain, whatever that means:
```
Computing //pc:peerconnection(//build/toolchain/win:x64)
```
and later compiled as:
```
[875/2936] CXX obj/pc/peerconnection/peerconnection.obj
```

Alternatively this could be a compiler bug.

**Comment by SergioOpenPeer** [comment]
@djee-ms Please, can you tell me what project are you running when you debugging? If you just added project in the WebRtc.Win32 solution, that references Org.WebRtc and trying to build for Any CPU, native lib and wrapper project will be built for x86. To build for x64, it is required to set it for build CPU, or modify default CPU in Configuration Manager . If it is mandatory for you to build using Any CPU, you have to use nuget package. Only that way it can be loaded proper lib.

**Comment by SergioOpenPeer** [comment]
PlatformTarget fix is pushed to releases/m71

**Comment by djee-ms** [comment]
To compile, in a first step I am using the `WebRtc.Win32` solution to build the `Org.WebRtc.dll` C# .NET Standard 2.0 wrapper and the supporting `libOrg.WebRtc.WrapperC.dll`. Then in a second step (different solution) I use them from a Unity script (.NET Standard backend) running inside the Unity Editor (x64 native process).

To debug, I attach Visual Studio to the Unity Editor [manually without using the VS Unity tool extension (to get native debugging)](https://forum.unity.com/threads/how-to-step-into-a-native-c-dll-in-visual-studio.413387/) and press the Play button in the Editor. This eventually calls the `Org.WebRtc` C# wrapper, which in turns call the `libOrg.WebRtc.WrapperC.dll` native code.

I already set the Visual Studio build config in both cases to Debug x64. I also made sure that `<PlatformTarget>` was set to x64 for `Org.WebRtc` by modifying the project properties manually. I don't think there's any "Any CPU" left anywhere. I also validated that the `Org.WebRtc.dll` library was compiled as x64 by opening the DLL with ILSpy.Net and check the assembly metadata (this was x86 before the `<PlatformTarget>` change, now is x64 as expected).

The code compiled for `webrtc.lib` reads the C++ structure field at different offset than MSVC compiler expects it to be when running. Although the `<PlatformTarget>` was an issue, I do not think this is enough to explain that some C++ native code from Google has some invalid offset. Can you please confirm whether `webrtc.lib` is compiled by Clang or MSVC when building `WebRtc.Win32.Native.Builder`? I suspect this is the former.

**Comment by SergioOpenPeer** [comment]
libyuv_internal library is built with clang-cl. Maybe that is causing an issue. I'll create a new branch, where libyuv_internal will be built using cl.

**Comment by djee-ms** [comment]
The problem seems to be around `absl::optional<>` vs. `std::optional<>`. In VS debugging, I get:
```cpp
sizeof(absl::optional<int>) | 12 | unsigned __int64
sizeof(std::optional<int>)  |  8 | unsigned __int64
```
but from the Google's point of view those should be equal, as `absl::optional` is just a replacement for when `std::optional` is not available. This introduces some offset into the `RTCConfiguration` structure.

**Comment by SergioOpenPeer** [comment]
From what you are saying, libyuv change won't resolve the issue. Anyway change is pushed to Sergej/20190509-net-fixes branch. Now clang-cl is not used for building any of the libraries. Let me know if this has changed anything.

**Comment by djee-ms** [comment]
Thanks for the quick update @SergioOpenPeer. Yes indeed, `libyuv` might pose some issue later but currently that's unrelated, I crash in the `RTCPeerConnection` constructor when parsing the `RTCConfiguration` input argument to its constructor, so very early in the WebRTC setup process.

I found that `webrtc.lib` is compiled by `cl.exe` from VS 2017 but without the `/std:c++17` flag, so Abseil cannot use `std::optional` and falls back to its own `absl::optional`. This is problematic for MSVC because it seems to add some extra padding at the end of the structure due to the 2 extra base classes without data (`optional_ctor_base` and `optional_assign_base`, see [here](https://github.com/abseil/abseil-cpp/blob/dbae8764fbd429bf7d7745e24bcf73962177a7c0/absl/types/optional.h)), something I think does not happen with Clang which probably see those as empty classes so optimizes them away in the structure layout.

So I am now trying to force `/std:c++17` from `webrtc\xplatform\chromium\build\toolchain\win\BUILD.gn` and hard-coding an `#define ABSL_HAVE_STD_OPTIONAL 1` inside `webrtc\xplatform\chromium\third_party\abseil-cpp\absl\types\optional.h` so that everything uses `std::optional`. This is not a good long-term solution, but as a temporary workaround if that works this would at least unblock me.

**Comment by djee-ms** [comment]
That seems to have worked, I crashed later in my `VideoCapturer.Create()` call, which is a few lines past the `PeerConnection` constructor.

So... can we consider the option to compile with `/std:c++17` and/or investigate how to fix `absl::optional` such that MSVC 14.0 (VS 2017) computes for it the same size as `std::optional`?

**Comment by SergioOpenPeer** [comment]
@djee-ms Did you just add /std:c++17, or you also added #define ABSL_HAVE_STD_OPTIONAL 1

**Comment by djee-ms** [comment]
I have 2 local changes:

1. `webrtc\xplatform\chromium\third_party\abseil-cpp\absl\types\optional.h`

```cpp
#ifndef ABSL_TYPES_OPTIONAL_H_
#define ABSL_TYPES_OPTIONAL_H_

#include "absl/base/config.h"
#include "absl/utility/utility.h"

#define ABSL_HAVE_STD_OPTIONAL 1 //<<<<<< Added this line 
#ifdef ABSL_HAVE_STD_OPTIONAL

#include <optional>
```

2. `webrtc\xplatform\chromium\build\toolchain\win\BUILD.gn`
Change the **6** instances of `command=` in that file to add the `/std:c++17` flag. Also add `/FS` because of some multi-threaded PDB issue during compiling, but I do not know if this is needed in general.

```cpp
command = "$env_wrapper$cl /nologo /showIncludes /FS /std:c++17 ${clflags} @$rspfile ...
```

**Comment by SergioOpenPeer** [comment]
Invalid build config for Org.WebRtc.csproj is fixed, c++17 is resolved under another ticket
